js/controller.js

```diff
@@ -687,15 +698,55 @@ OO.plugin("Html5Skin", function (OO, _, $, W) {
      ---------------------------------------------------------------------*/
     toggleFullscreen: function() {
       this.state.fullscreen = !this.state.fullscreen;
-      if(this.state.fullscreen) {
-        this.state.mainVideoWrapper.addClass('fullscreen');
+      if(this.state.isFullScreenSupported) {
+        Fullscreen.toggle(this.state.mainVideoWrapper.get(0));
       } else {
-        this.state.mainVideoWrapper.removeClass('fullscreen');
+        if(this.state.isFullWindow) {
+          this.exitFullWindow();
+        } else {
+          this.enterFullWindow();
+        }
       }
-      this.mb.publish(OO.EVENTS.WILL_CHANGE_FULLSCREEN, this.state.fullscreen);
       this.renderSkin();
     },
 
+    // if fullscreen is not supported natively, "full window" style
+    // is applied to video wrapper to fill browser window
+    enterFullWindow: function() {
+      this.state.isFullWindow = this.state.fullscreen = true;
+
+      // add listener for esc key
+      document.addEventListener("keydown", this.exitFullWindowOnEscKey.bind(this));
+
+      // hide scroll bars
+      document.documentElement.style.overflow = 'hidden';
+
+      //apply full window style
+      this.state.mainVideoWrapper.addClass('fullscreen');
+    },
+
+    // remove "full window" style and event listener
+    exitFullWindow: function() {
+      this.state.isFullWindow = this.state.fullscreen = false;
+
+      // remove event listener
+      document.removeEventListener("keydown", this.exitFullWindowOnEscKey);
+
+      // unhide scroll bars
+      document.documentElement.style.overflow = 'visible';
+
+      //remove full window style
+      this.state.mainVideoWrapper.removeClass('fullscreen');
+    },
+
+    // exit full window on ESC key
+    exitFullWindowOnEscKey: function(event) {
```

Comment:
This exit with ESC button is only applied to fullWindow. Does the other true fullScreen already has ESC button functionality?